### ISSUE TITLE
Bug: @typescript-eslint/no-unused-vars False Positives on Interfaces and types

### ISSUE THREAD

**Issue by agarwalvaibhav0211** [typescript-eslint/typescript-eslint#7218]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.5.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgIIEZkG8BQzlwBcyAzmFKAOY4C%2BOYAngA4qoBMyAvNngcWRRDUaAbhwQQAVwC2aAMzZkNIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYwAvgDThTJHSoCuAJuhWANoMSQAdAHoAVKLGC8ZBgF0Q1IA&tokens=false

### Repro Code

```typescript
interface A1 {
  a:string
}
type A2 = {
  a: string
};
enum A3 {}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {},
  "include": [
    "./**/*.ts"
  ]
}
```

### Expected Result

There should be no errors as the types and interfaces can be used in other files.

### Actual Result

there are errors

```
'A1' is defined but never used. 1:11 - 1:13
'A2' is defined but never used. 4:6 - 4:8
'A3' is defined but never used. 7:6 - 7:8
```
![image](https://github.com/typescript-eslint/typescript-eslint/assets/43950046/d1688d00-38c1-404f-88d4-42570e4bf5da)

**Comment by bradzacher** [comment]
> as the types and interfaces can be used in other files

Not if you're using modules, they can't.
If you're not using modules and you're leveraging another method for globally sharing things across `.ts` files - the docs describe how you can configure this.

https://eslint.org/docs/latest/rules/no-unused-vars#exported

This is a generally unsupported workflow though because global usage is strongly discouraged and instead explicit exports with modules is the baseline practice.
